dist/tools/testrunner/testrunner.py

```diff
@@ -7,32 +7,389 @@
+
+def distribute(possible, _all=False):
+    for test, node_groups in possible.items():
+        if not node_groups:
+            continue
+
+        if _all:
+            test.node_groups = node_groups
+        else:
+            test.node_groups = [sorted(node_groups, key=lambda x: len(x.tests))[0]]
+
+        for node_group in test.node_groups:
+            node_group.tests.append(test)
+
+def check_results(results):
+    if results["result"]:
+        return False
+
+    terms = listify(results.get("check_result", []))
+    if not terms:
+        return True
+    else:
+        print("testrunner: checking results")
+
+        res = False
+        _check_results = {}
+        for term in terms:
+            _res = eval(term, {}, results["results"])
+            res = res or bool(_res)
+            print("testrunner: '%s'==%s (%s)" % (term, res, bool(res)))
+            _check_results[term] = _res
+
+        print("testrunner: %s" % "passed" if res else "failed")
+        results["check_result"] = _check_results
+        return res
+
+def run_tests(tests, outfile, instance):
+    exit_code = 0
+    results = []
+    for test in tests:
+        if test.instance_id != instance:
+            continue
+        if not test.node_groups:
+            print("%s: no matching group found!" % test.filename, file=sys.stderr)
+            continue
+        for node_group in test.node_groups:
+            print("- running test", test.filename, "on node group", node_group.filename, end="")
+
+            try:
+                print(" (instance %s: %s)" % ( test.instance_id, test.data.get("template_instance", {})))
+            except KeyError:
+                print("")
+
+            test.test_class = _tests.map.get(test.data["test"])
+            test_nodes = node_group.create_node_objs(test)
+
+            test_obj = test.test_class(test_nodes, args=test.data.get("args", {}))
+
+            # run test
+            try:
+                test_obj.safe_run()
+            except KeyboardInterrupt as e:
+                if not is_main_interrupted():
+                    raise e
+                else:
+                    test_obj.result = "error"
+
+            result = {}
+
+            node_data = []
+            for node in test_nodes:
+                node_data.append(node.data)
+
+            result["used_nodes"] = node_data
+            result.update(test.data)
+            result.update(test_obj.result_data or {})
+            result["result"] = test_obj.result
+
+            test_obj.result = test_obj.result or (not check_results(result))
+            result["output"] = test_obj.output
+
+            if test_obj.result:
+                print("- TEST FAILED!")
+                exit_code = 1
+            else:
+                print("- TEST SUCCESSFUL!")
+
+            results.append(result)
+
+        outfile.write(json.dumps(results, indent=4, sort_keys=True, default=set_default))
+
+    return exit_code
+
+def print_murdock_jobs(tests):
```

Comment:
I'm getting some stomach ache seeing the name Murdock in something that doesn't have anything to do with Murdock (except that it is supposed to run in it).